testing/webutils.rb

```diff
@@ -0,0 +1,331 @@
+# frozen_string_literal: true
+
+require_relative 'config.rb'
+
+def each_browser(p)
+  TESTCONFIG[:browsers].each do |b|
+    p.call(b)
+  end
+end
+
+module WebTestUtils
+
+  def set_base_url(u)
+    @baseurl = u
+  end
+
+  def start_server(browser, headless = false)
+    if (browser == :firefox) && headless
+      options = Selenium::WebDriver::Firefox::Options.new(args: ['-headless'])
+      @browser = Selenium::WebDriver.for browser, options: options
+    elsif (browser == :chrome) && headless
+      options = Selenium::WebDriver::Chrome::Options.new(args: ['headless'])
+      @browser = Selenium::WebDriver.for browser, options: options
+    else
+      @browser = Selenium::WebDriver.for browser
+    end
+    @wait = Selenium::WebDriver::Wait.new(:timeout => 10)
+    @longwait = Selenium::WebDriver::Wait.new(:timeout => 30)
+  end
+
+  def shutdown_server
+    @browser.quit
+  end
+
+  def nav_to_home(browser = @browser)
+    puts 'in nav_to_home'
+    x = browser.find_element(:tag_name, 'html')
+    puts "got old page element: #{x}"
+    browser.get "#{@baseurl}/"
+    puts 'waiting on home page load'
+    @wait.until do
+      puts '...polling'
+      y = browser.find_element(:tag_name, 'html')
+      puts "...got element #{y}"
+      x != y
+    end
+    puts 'done with home page load'
+  end
+
+  def nav_to(path, browser = @browser)
+    puts "in nav_to: #{path}"
+    x = browser.find_element(:tag_name, 'html')
+    puts "got old page element: #{x}"
+    browser.get "#{@baseurl}/#{path}"
+    puts 'waiting on page load'
+    @wait.until do
+      puts '...polling'
+      y = browser.find_element(:tag_name, 'html')
```

Comment:
Could you add a brief comment (for my sake) about how this works? Especially given the more convoluted "wait for stale handle" method below, I'm curious to know how this is working

Reply:
Added... though probably not entirely satisfactory explanation